File: lib/model/access_request.ts
import { Knex } from 'knex';

export interface AccessRequest {
  id: string;
  userId: string;
  key: string;
  granted?: boolean;
  decisionTimestamp?: Date;
}

const insert = (db: Knex) => (accessRequest: AccessRequest) =>
  db
    .insert({ ...accessRequest, createdAt: new Date() })
    .into('accessRequest')
    .returning('*')
    .then((r) => r[0]);

const upsert = (db: Knex) => (accessRequest: AccessRequest) =>
  db
    .insert({ ...accessRequest, createdAt: new Date() })
    .into('accessRequest')
    .onConflict('id')
    .merge({ ...accessRequest, updatedAt: new Date() })
    .returning('*')
    .then((r) => r[0]);

const get =
  (db: Knex) =>
  (id: string): Promise<AccessRequest> =>
    db
      .select()
      .from('accessRequest')
      .where('id', id)
      .limit(1)
      .then((r) => (r.length ? r[0] : null));

const getByUserId =
  (db: Knex) =>
  (userId: string): Promise<AccessRequest[]> =>
    db
      .select()
      .from('accessRequest')
      .where('userId', userId)
      .orderBy('createdAt', 'asc');

const getByUserIdAndKey =
  (db: Knex) =>
  (userId: string, key: string): Promise<AccessRequest> =>
    db
      .select()
      .from('accessRequest')
      .where({ userId, key })
      .limit(1)
      .then((r) => (r.length ? r[0] : null));

const getAll = (db: Knex) => () =>
  db.select().from('accessRequest').orderBy('createdAt', 'asc');

const del = (db: Knex) => (id: string) => db.del().where('id', id);

export default (db: Knex) => ({
  insert: insert(db),
  upsert: upsert(db),
  get: get(db),
  getByUserId: getByUserId(db),
  getByUserIdAndKey: getByUserIdAndKey(db),
  getAll: getAll(db),
  del: del(db),
});

File: src/models/access_request.rs
use anyhow::Result;
use chrono::NaiveDateTime;
use serde::Serialize;
use sqlx::PgPool;
use uuid::Uuid;

#[derive(Debug, Serialize, sqlx::FromRow)]
#[serde(rename_all = "camelCase")]
pub struct AccessRequest {
    pub id: Uuid,
    pub user_id: Uuid,
    pub key: String,
    pub granted: Option<bool>,
    pub decided_at: Option<NaiveDateTime>,
    pub granted_from: Option<NaiveDateTime>,
    pub granted_until: Option<NaiveDateTime>,
    pub created_at: Option<NaiveDateTime>,
    pub updated_at: Option<NaiveDateTime>,
}

pub async fn insert(pool: &PgPool, user_id: Uuid, key: &str) -> Result<AccessRequest> {
    let now = chrono::Utc::now().naive_utc();
    let ar = sqlx::query_as::<_, AccessRequest>(
        "INSERT INTO access_request
             (id, user_id, key, granted, decided_at, granted_from, granted_until, created_at)
         VALUES ($1, $2, $3, NULL, NULL, NULL, NULL, $4)
         RETURNING *",
    )
    .bind(Uuid::new_v4())
    .bind(user_id)
    .bind(key)
    .bind(now)
    .fetch_one(pool)
    .await?;
    Ok(ar)
}

pub async fn get(pool: &PgPool, id: Uuid) -> Result<Option<AccessRequest>> {
    let row = sqlx::query_as::<_, AccessRequest>(
        "SELECT * FROM access_request WHERE id = $1",
    )
    .bind(id)
    .fetch_optional(pool)
    .await?;
    Ok(row)
}

pub async fn delete(pool: &PgPool, id: Uuid) -> Result<()> {
    sqlx::query("DELETE FROM access_request WHERE id = $1")
        .bind(id)
        .execute(pool)
        .await?;
    Ok(())
}

pub async fn get_by_user_id_and_key(
    pool: &PgPool,
    user_id: Uuid,
    key: &str,
) -> Result<Vec<AccessRequest>> {
    let rows = sqlx::query_as::<_, AccessRequest>(
        "SELECT * FROM access_request WHERE user_id = $1 AND key = $2 ORDER BY created_at DESC",
    )
    .bind(user_id)
    .bind(key)
    .fetch_all(pool)
    .await?;
    Ok(rows)
}
